### ISSUE TITLE
Add a feature to pull unit_count for filters from index name

### ISSUE THREAD

**Issue by soenkeliebau** [elastic/curator#996]
We are currently looking into building a flexible log management architecture that gives the log sender some control over how his log events are treated by processing metadata about that log source.
One of those fields is a time to live in days, so how long that event should be kept, before it is deleted.

Events are then stored in indices called logstash-30-yyyy.mm.dd (for a 30 day ttl) or logstash-60-yyyy.mm.dd (for a 60 day ttl) etc.
In order to enable housekeeping on these we'd need to create an action in curator per existing value for ttl (unless I am missing something), which is not desireable.
Would it be of interest to add functionality to specify a regex or something similar to enable curator to dynamically pull the value for unit_count from the index name? I'd be happy to look into this and come up with a proposal of how configuration could look first and work on a PR later on if people think this makes sense.

**Comment by untergeek** [comment]
It's a fascinating idea.  I cannot think that this could hope to work without a regular expression with a captured group, though.  That's certainly not a deal breaker, but it would be daunting for those not familiar with regular expressions.

The larger question for me is centered around the rollover API and action.  Elasticsearch is headed towards recommending _away_ from named indices in favor of using the rollover API.  While you clearly could name indices with a numeric value and operate in a similar way, it becomes less important to do so, as you end up rotating based on document count and/or age relative to creation_date, rather than index name.  Does this approach still offer value if you're doing that?  It could, I guess, but with such a rotation, it seems it would be more valuable to use the field_stats API (or the new API replacing it) to determine the age of the oldest and newest documents, so as to not delete anything not _fully_ out of the retention period.

**Comment by soenkeliebau** [comment]
I agree with you on the regex that would be necessary for this to work. While in theory we could probably come up with a defined marker for the unit count and a syntax similar to the index name pattern the longer I thought about it the more complex I had to make it to accomodate the cases I came up with. Regexes have been around for a long time, are extremely powerful and well documented (as in there is a lof of documentation, not necessarily saying all of it is good) - so no need to reinvent the wheel.
I won't pretend to know the background of the average curator user, but I think a fair few of them have grep-ed some log files back in the day and used regular expressions to do so, so I am confident that this should work out nicely.

On your second question, I think that this will play quite nicely with the rollover feature, since the underlying principle is the same.
For example, if we have two different retention times, 30 and 60 days, if I don't misunderstand the rollover pattern, the indices for this might look like this:

logs-30-1
logs-30-2
logs-30-3
logs-60-1
logs-60-2
logs-60-3

If we tell curator to delete indices containing data older than the retention time and use index names to determine calculation times, we'd probably have a regex like this: "logs-([0-9]+)-[0-9]+". Apply this to the index names and extract 30 or 60 and store this as unit_count per index. Processing after this doesn't need to change much I think.


I've looked through the code a bit and unless I totally miss my mark (which I probably am), I don't think that this would be too large of a change to implement.
My idea was to add a new parameter to the age filter "unit_count_pattern" which stores the regex. Maybe change the parameters so that at least one of "unit_count" or "unit_count" pattern has to be present, but both are allowed too. If both are there we can fall back to the given unit count if the pattern doesn't match, if only the pattern is present indices that don't match the pattern are ignored.
The actual implementation would I think go [here](https://github.com/elastic/curator/blob/master/curator/indexlist.py#L448). Check if a pattern was defined, if yes compare it to the index name, extract the retention time and change the "agetest" variable to compare against a different point of reference.

There are probably a whole lot of issues that would creep up while implementing this, but in principle I think it could work.

**Comment by untergeek** [comment]
If you want to tackle this, I'm all for it!  The planned behavior seems sound, so let's be sure to test it with unit tests and actual integration tests.  I'm happy to help with those.

**Comment by soenkeliebau** [comment]
Right, I'll give it a shot. I am sure I'll need some help with tests though, thanks for offering!
Quick question up front, is there an existing way to link two filter elements so that one of them has to be present? If not, I'd probably rethink the way to configure this to avoid using a dependency like that.

**Comment by soenkeliebau** [comment]
I've created a pull request with some code purely for a high level review of the approach taken. This does work for the tests that I performed and doesn't seem to break any unit tests. I had a few failures when running them locally, but those were about snapshots which my ES instance is probably not correctly set up for.
I've changed the parameter behavior for now as follows:
unit_count remains mandatory
unit_count_pattern is optional, if it is present, it is matched against the index name, if a successful match is returned that is a number that number is used instead of unit_count - if a match is not possible, unit_count is used as fallback, unless it is set to -1 in which case the index is failed.
This is probably not how we want to keep, but was easiest for me to implement for intial testing.
